2764 Adds list of synonyms and stopwords to OA elasticsearch fields

albertisfu opened the pull request:
In this PR I added a custom analyzer in order to emulate the one existing in Solr [text_en_splitting_cl](https://github.com/freelawproject/courtlistener-solr-server/blame/main/solr/cores/audio/conf/schema.xml#L529) used in the following Oral Arguments fields:

- caseName
- court
- docketNumber
- judge
- text 

This analyzer includes the following characteristics:
Tokenizer: 
solr: `WhitespaceTokenizerFactory` -> ES: `whitespace`
Filters:
solr: `StopFilterFactory` -> ES: `english_stop`
solr: `WordDelimiterFilterFactory` -> ES: `word_delimiter`
solr: `PatternReplaceFilterFactory` -> ES: `remove_leading_zeros`
solr: `LowerCaseFilterFactory` -> ES: `lowercase`
solr: `SynonymFilterFactory` -> ES: `synonym_filter`
solr: `KeywordMarkerFilterFactory` -> ES: `keyword_marker`
solr: `SnowballPorterFilterFactory` -> ES: `english_stemmer`
solr: `RemoveDuplicatesTokenFilterFactory`  -> ES:  `remove_duplicates` 

For development and testing purposes I added the `synonyms_en.txt` `stopwords_en.txt` `protwords_en.txt` files and mounted them in the `cl-es` docker-compose service. 

Only the `synonyms_en.txt` file required some syntax tweaks to be compatible with Elasticsearch.

@DictCheney would you mind giving it a look and seeing how this solr-elastic analyzer migration looks to you?
Solr schema link: [text_en_splitting_cl](https://github.com/freelawproject/courtlistener-solr-server/blame/main/solr/cores/audio/conf/schema.xml#L529)
Elastic analyzer definition:  [text_en_splitting_cl ](https://github.com/freelawproject/courtlistener/pull/2765/files#diff-8a24d083126e5cee1f14a9d77ee1bdd0a238cd158f0ad0e9e3153b226ed57879)

- I added new tests to ensure that the synonyms and stopwords search functions correctly.
- Additionally, I included a test to verify that the bug reported in https://github.com/freelawproject/courtlistener/issues/1635 does not exist in Elasticsearch after incorporating synonyms and stopwords.
- I made some adjustments to existing tests due to the newly added fixtures and tests related to relevance sorting. After implementing the new analyzer, the order of results is now similar to that of Solr. In cases where the results matched the same keywords and the order appeared arbitrary due to the same score, the new order is more consistent.

albertisfu commented:
@mlissner Per our Slack conversation, the `stopwords` filter is now removed from the analyzer.

Before searches like:
`Wallace to Friedland` or `Wallace and Friedland` or `"Wallace to Friedland"` or `"Wallace and Friedland"` returned results for both  `Wallace to Friedland` and `Wallace and Friedland` 

Now after removing the stopwords filter stopwords are considered in the search so that queries like:
`Wallace to Friedland` only returns results for `Wallace to Friedland`
`Wallace and Friedland` only returns results for `Wallace and Friedland`

- Added tests for these cases works properly.

![Screenshot 2023-05-18 at 16 32 44](https://github.com/freelawproject/courtlistener/assets/486004/2385cd43-e007-4e49-9064-2bdad4b13fdf)

I also checked [Solr tokenizer breaks up by character case #824](https://github.com/freelawproject/courtlistener/issues/824).

e.g., queries like `WikiLeaks` or `GraphQL` returns different results than `Wikileaks` or `Graphql` even though they're the same words, this issue is not present in Elasticsearch with the current analyzer.

However, I have identified the problem in Solr.

In Elasticsearch, the order in which filters are applied is:

```
"lowercase"
"synonym_filter"
"custom_word_delimiter_filter"
"remove_leading_zeros"
"keyword_marker"
"english_stemmer"
"remove_duplicates"
```

In Solr, the word_delimiter_filter is applied before the lowercase filter. This means that strings like `WikiLeaks` are split into tokens like `Wiki` and `Leaks` first, and then lowercased to `wik`i and `leaks`. As a result, the search is based on these tokens. On the other hand, if the string is `Wikileaks` it is not split, so the final string to search remains as `wikileaks`.

So the solution, in this case, is to apply first the `lowercase` filter and then apply the `word_delimiter` filter, so camel-cased strings are not affected by the filter.

- Also added tests to confirm this issue is also solved in ES.

The `word_delimiter` filter is still needed so another type of string like dash separated e.g: `Lynch-Loretta` can be also found by separated tokens like `lynch` `loretta`.

I just tweaked the `word_delimiter` to disable the `split_on_numerics` param which is also disabled in Solr (which prevents split strings like 3D into `3` and `d`)
All the other custom params are the same as in Solr since are default values.

I'll continue checking other search-related issues like #1645 so the analyzer might still need some tweaks.

mlissner commented on cl/search/elasticsearch_files/protwords_en.txt:
Seems like we're not using this and may as well remove it. If we haven't needed it yet.....we probably don't need to carry it over to ES.

albertisfu replied:
Yeah, I've removed the `keyword_marker` filter (that used this file) from the analyzer and also removed the `protwords_en.txt` file.

albertisfu commented:
Suggestions applied, let me know what you think about stopwords.
Thank you!

albertisfu commented on cl/search/tests.py:
```diff
@@ -3111,6 +3126,41 @@ def test_oa_case_name_filtering(self) -> None:
         )
 
+    def test_oa_docket_number_filtering(self) -> None:
+        """Filter by case_name"""
```
Changed!

albertisfu commented on cl/search/templates/includes/search_result.html:
```diff
@@ -37,7 +37,7 @@ <h3 class="bottom serif">
          class="visitable">
       {{ result.caseName|safe  }}
-      ({% if result.court_id != 'scotus' %}{{ result.citation_string|nbsp|safe }}&nbsp;{% endif %}{{ result.dateArgued|date:"Y" }})
+      ({% if result.court_id != 'scotus' %}{{ result.court_citation_string|safe }}&nbsp;{% endif %}{{ result.dateArgued|date:"Y" }})
```
Yeah, fixed it. It needed to tweak the `nbsp` template filter in order to avoid escaping the `<mark>` tags after marking them as safe.

albertisfu commented on cl/search/elasticsearch_files/stopwords_en.txt:
Yeah, we decided to not use `stopwords` in the custom analyzer. We're not using it:

```
"filter": [
                    "lowercase",
                    "synonym_filter",
                    "custom_word_delimiter_filter",
                    "remove_leading_zeros",
                    "english_stemmer",
                    "remove_duplicates",
                ]
```

I just forgot to remove the stopwords_en.txt file.

So let me know if we should use stopwords for terms at the top with the X's and remove all the other words or if we should just remove the stopwords_en.txt file.

mlissner replied:
We need those two at the top. We have code that finds social security numbers and changes them to X's. If we don't have the X's stopwords, people could just search for the X'ed out docs and find everybody's SSN in the system.

albertisfu replied:
Done!

mlissner commented:
Back at you.

albertisfu commented:
I've kept only the special stopwords (with the X's) and added them to the analyzer.  Also added a test to confirm these words are not searchable.

Also updated this branch and solve merge conflicts.

Let me know if there is anything else here.

Thanks!

albertisfu commented:
@mlissner I've added the list of top abbreviations you built in #2791

I did some testing to be sure everything is working properly and I found some possible issues:

In #2791 you mentioned that you added synonyms using the format:
`convention against torture => cat`

According to your description, a search for` convention against torture` should return results for `cat`, which is working as expected. However, searching for `cat` did return documents containing `convention against torture`.

The way the `=>` seems to be working is the following:

`convention against torture => cat`

In this case, the three words 'convention against torture' are replaced with 'cat'.

So when searching for `convention against torture` it returns documents containing `cat` but also doing a search for `cat` is returning documents containing `convention against torture`, seems the analyzer is treating as equivalent both sides of the` =>`.

The difference between the examples` administrative procedures act => apa` and `administrative procedures act, apa` is that the `=>` definition considers the entire phrase as a single term, whereas `administrative procedures act, apa` breaks `administrative procedures act` into individual tokens.

So using the `administrative procedures act, apa` definition, if you look for `administrative procedures act` it'll return results for `apa` and vice versa.

However, the problem arises when searching for just `administrative`. It will return results containing `apa`, as well as other terms containing `administrative`, such as `oah` (office of administrative hearings). Is that correct?

If we want to only return results containing `apa` when the whole phrase `administrative procedures act` is matched, we should use the `=>` definition.

Please let me know which matching behavior we want, and I can make the necessary adjustments to the synonyms file.

mlissner commented:
Yeah, your understanding of the intention is correct:

 - When the acronym is a common word (like `cat`), we want to search for the phrase and get results with it and the acronym.

    I thought this was handled by `convention against torture => cat`, but apparently not.

 - When the acronym is not a common word (like `FOIA`), searches for the phrase or the acronym should be treated equally.

    I thought this was handled by comma separated values.

I saw some of the existing synonyms use double quotes around phrases. Would that help?

albertisfu commented:
Understood. Yes, it seems that in Solr, double quotes are used to preserve synonyms as phrases. However, in Elasticsearch it breaks the indexing due to offset differences during the analyzer transformation. 

The way to do it appears to be using the` =>` notation. But I'll see how to fix common words synonyms don't return results related to the acronym.

albertisfu commented:
I reviewed synonyms functioning in-depth and did some experimenting in order to achieve the synonyms' behavior that we expect.

These are my findings:

- I confirmed one-way synonyms are not possible to implement using:
`convention against torture => cat`

When using this syntax the indexing occurs as follows: whenever is found `convention against torture` it’s indexed as `cat` (type synonym) 

If the document contains `cat` it’s indexed as `cat` (type word)

So when performing a search, if the user looks for `convention against torture` it’s transformed into `cat` in search time, so the term `cat` is used to find matches in the index.

So that `convention against torture` will return results for documents containing: `cat`

If the user search for `cat` not transformation is performed, so it’ll match documents containing `cat` and `convention against torture` (indexed as `cat`).
It seems there is no straightforward way to achieve one-way synonyms in elastic search.

- About the other issue related to match synonyms as a phrase.

Using the approach in this PR of using explicit mappings solves the issue.

Syntax:
`administrative procedure act, administrative procedures act => apa`

In this case, documents containing `administrative procedure act` or `administrative procedures act` will be indexed as `apa`

So searching for `apa` will return results for `administrative procedure act` or `administrative procedures act`.
Also, searching for `administrative procedures act` will return results for `apa`, `administrative procedures act` or `administrative procedure act`, which is correct.

_But a search for `administrative` won’t return results for `apa` (which is correct). But also won’t return results for documents containing `administrative procedure act`, which means the synonym can't be partially matched, since these synonyms terms are indexed as `apa`_ 

- My question here is if this behavior is correct.

_Or if a search for `administrative` should return results for documents containing `administrative` as part of a synonym like `administrative` procedure act while avoiding show results for `apa`?_
![Screenshot 2023-06-06 at 17 20 44](https://github.com/freelawproject/courtlistener/assets/486004/19fa85aa-ab38-43be-9056-4d92e2fa5ca5)

Because this behavior is also possible, I found a way to achieve these results using the expanding syntax, using a different analyzer in search time, and removing the synonyms filter for the index time analyzer.

@mlissner  please just let me know which one is the approach we expect.

In addition, It seems to have some advantages avoiding indexing synonyms and just using synonyms on search time:

https://www.elastic.co/blog/boosting-the-power-of-elasticsearch-with-synonyms

> Index-time synonyms have several disadvantages:
> 
> - The index might get bigger, because all synonyms must be indexed.
> - Search scoring, which relies on term statistics, might suffer because synonyms are also counted, and the statistics for less common words become skewed.
> - Synonym rules can’t be changed for existing documents without reindexing.
> 
> Using synonyms in search-time analyzers on the other hand doesn’t have many of the above mentioned problems:
> 
> - The index size is unaffected.
> - The term statistics in the corpus stay the same.
> - Changes in the synonym rules don’t require reindexing of documents.

Finally, I’m afraid we won’t be able to match synonyms for hypenated terms like `post-traumatic stress disorder` since the hyphenated terms are messing up the original positions in terms. So during search time terms are not properly matched.

But in this case, it’ll still work for `post traumatic stress disorder, posttraumatic stress disorder, ptsd`

Also if the document indexed is `post-traumatic stress disorder` we'll be able to return results if the search is for any of the synonyms since during index time the `-` is removed.

The only exception is when a search for `post-traumatic stress disorder` won't return documents for its synonyms.

mlissner commented:
>  It seems there is no straightforward way to achieve one-way synonyms in elastic search.

That's a bummer. So if this is true, we should exclude any synonyms where the abbreviation is a common English word.

> a search for administrative won’t return results for apa (which is correct). But also won’t return results for documents containing administrative procedure act, which means the synonym can't be partially matched, since these synonyms terms are indexed as apa

That's pretty lame, hard to explain to users, and weird. 

>  I found a way to achieve these results using the expanding syntax, using a different analyzer in search time, and removing the synonyms filter for the index time analyzer.

This sounds promising to me. Index-time synonyms _are_ frustrating since they can't be updated without reindexing millions of documents. Does this fix directional synonyms too? 

> The only exception is when a search for post-traumatic stress disorder won't return documents for its synonyms.

So be it. There are only a few hyphenated phrases. We can survive without expanding them into synonyms. 

Thanks for the analysis. Do you have what you need to proceed here?

albertisfu commented:
Thanks!

About one-way synonyms, I tried an additional approach described [here](https://stackoverflow.com/questions/34333486/one-way-synonym-search-in-elasticsearch) inverting the definition of the synonym.
`apa => procedures act, administrative procedure act`

And it works for one-way synonyms, however, there are a couple of issues:
- It works on index-time. 
- Partial synonym word matches also return results for the acronym. For example, when searching for "administration," results for "apa" are also included, which is not correct.

It seems that the [only reliable way](https://stackoverflow.com/a/34336575) to achieve one-way synonyms is not to use ES synonyms filter and instead, perform a custom pre-processing before indexing to ES, looking for synonyms occurrences and replacing them with their synonyms according to the behavior we want, e.g when `convention against torture` is found we replace it with `cat` but if `cat` is in the original text it remains the same.  Each synonym version should be stored in a different field to avoid affecting scores. Which seems complex and will make the index bigger.

So yeah, I think the best approach is to remove synonyms where the abbreviation is a common English word.

> This sounds promising to me. Index-time synonyms are frustrating since they can't be updated without reindexing millions of documents. Does this fix directional synonyms too?

Yeah, and it's an easy change, however, it doesn't solve the one-way synonyms either.

> Do you have what you need to proceed here?
Yes, I'll implement synonyms only in search time, so this will also solve the problem related to synonyms can't be partially matched.

Also, if it's ok, I'll remove synonyms where the abbreviation is a common English word.

mlissner commented:
Sounds great, thank you.

> Also, if it's ok, I'll remove synonyms where the abbreviation is a common English word.

Yep!

albertisfu commented:
I've applied the new approach of only handling synonyms at search time and removing them from indexing time.

This also resolves the issue of partially matching multi-word synonyms. So a search for `administration` will return results for documents containing "administrative procedures act" while keeping previous functionalities:

Search for `apa` returns results for  `administrative procedures act` and `apa`
Search for `administrative procedures act` returns results for `apa` and `administrative procedures act`

Also tweaked the list of  single word one-way synonyms:
`mag => mag,magazine,magistrate`

This way a search for `mag` will return results documents containing `mag`,`magazine`, or `magistrate`.
While a search for `magazine` will only return results for documents containing `magazine`.

- Added tests for all of these cases.

So I think this is ready, let me know if there is anything else here.

mlissner commented:
THIS IS READY TO MERGE.
